### PR TITLE
Geom: Integrate new Shape Hierarchy

### LINKED ISSUE #1122: Geom: Swap old shapes with new Shape Hierarchy
## Description of the task
Replace usages of old rectangle, polygon and triangle with new rectangle, polygon and triangle in new_geom. 

## Acceptance criteria
- [ ] Old geom shapes removed
- [ ] All usages of old shapes replaced with new geom implementation
- [ ] All unit tests still pass

### PR DESCRIPTION
**Pull request by scveloso**
### Description
Removes old geom shapes. Replaces their usages with new geom shape hierarchy. Adds new geom ```Rectangle::expand(...)```

### Testing Done
- Existing unit tests still pass
- ```Rectangle::expand(...)``` unit tests

### Resolved Issues
resolves #1122

### Length Justification
Most of the length is because of the removal of old geom polygon and rectangle. Other changes are replacements of usages from old geom to new geom. 

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/software/new_geom/convex_polygon.cpp
Comment (scveloso) [inline src/software/new_geom/convex_polygon.cpp]: ```GeomConstants::EPSILON``` is not tolerant enough for some convex polygons. For example, ```Triangle((Point(0, 0), Point(4.5, 0.5), Point(4.5, -0.5))``` should be a valid convex polygon but it is failing ```isConvex(...)``` check with our current EPSILON, I changed it to ```1e-12``` for now but I'm not sure how we should go about this? Is a different ```EPSILON``` for this okay?

Comment (garethellis0) [reply src/software/new_geom/convex_polygon.cpp]: So looking at the implementation of `isConvex`, I'm not super surprised we hit this issue. On each corner angle calculation we're going to accumulate some floating point error, which could easily make the final check fail. I think we probably want to change over to a numerically stable implementation. 

I'm fairly sure just checking that:
1) abs(each corner) is < 180 deg
2) All angles are either greater or less then zero (depending if points are given in CW or CCW order) ( _do we have tests for CW and CCW?_)
3) The polygon is not self-intersecting (this effectively replaces the final check)

File: src/software/ai/evaluation/indirect_chip.cpp
Comment (garethellis0) [inline src/software/ai/evaluation/indirect_chip.cpp]: Not a blocking comment, very much out-of-scope, but how would people feel about `mean` instead of `center` for a polygon?

File: src/software/ai/hl/stp/action/chip_action.cpp
Comment (garethellis0) [inline src/software/ai/hl/stp/action/chip_action.cpp]: TODO?

File: src/software/geom/util.cpp
Comment (garethellis0) [inline src/software/geom/util.cpp]: Seems like github has the diff kinda funky here. Could you merge `master` into your branch? Hopefully then most of this bit of the diff should vanish. Otherwise, we might need to remove some stuff here, @EvanMorcom rewrote a lot of this.

File: src/software/new_geom/rectangle.h
Comment (garethellis0) [inline src/software/new_geom/rectangle.h]: Could we specify the expected behavior if the rectangle can't shrink by the given amount? Will it shrink to a point or remain unchanged (or something else entirely)?

File: src/software/new_geom/util/intersects.cpp
Comment (garethellis0) [inline src/software/new_geom/util/intersects.cpp:19]: Where do the implementations for the `intersects` methods we're using here live?

**Review by garethellis0 — CHANGES_REQUESTED** [review]
Nice job on this, very happy with so many of these changes. :smile:

File: src/software/new_geom/convex_polygon.cpp
Comment (scveloso) [reply src/software/new_geom/convex_polygon.cpp]: I'll try changing the implementation of ```isConvex``` and see how that goes

File: src/software/geom/util.cpp
Comment (scveloso) [reply src/software/geom/util.cpp]: I rebased on master, Evan's changes are in the commit log for this branch. But if something is weird, let me know what I should remove

File: src/software/new_geom/rectangle.h
Comment (scveloso) [reply src/software/new_geom/rectangle.h]: It will remain unchanged but yes I'll add that in the comment

File: src/software/new_geom/util/intersects.cpp
Comment (scveloso) [reply src/software/new_geom/util/intersects.cpp:19]: It lives in geom/util.h and geom/util.cpp

File: src/software/new_geom/convex_polygon.cpp
Comment (scveloso) [reply src/software/new_geom/convex_polygon.cpp]: "The polygon is not self-intersecting"

Looked into this and it isn't trivial to implement. What I did for now is to change any use of degrees in ```isConvex``` to radians and it seems to be more precise as I was able to lower the EPSILON used. Maybe in a different issue, we can have a proper implementation of ```isConvex``` that uses what you mentioned above?

File: src/software/ai/hl/stp/action/chip_action.cpp
Comment (scveloso) [reply src/software/ai/hl/stp/action/chip_action.cpp]: Fixed in latest update

File: src/software/new_geom/rectangle.h
Comment (scveloso) [reply src/software/new_geom/rectangle.h]: Fixed in latest

File: src/software/geom/util.cpp
Comment (scveloso) [reply src/software/geom/util.cpp]: I looked at Evan's changes and it seems like my merge reintroduced ```angleSweepCircles(...)``` and its helper. I removed these in the latest update.

File: src/software/ai/evaluation/indirect_chip.cpp
Comment (MathewMacDougall) [reply src/software/ai/evaluation/indirect_chip.cpp]: I'm indifferent

**Review by MathewMacDougall — COMMENTED** [review]
Awesome stuff! Really glad to see _real_ geometry types instead of things like `LegacyTriangle` :fearful: . Nothing to add beyond Gareth's comments atm.

File: src/software/new_geom/convex_polygon.cpp
Comment (garethellis0) [reply src/software/new_geom/convex_polygon.cpp]: I'm totally fine if we open an issue here and put it off (as long as we prioritize it), but out of curiosity, what is the complexity with self intersections? Wouldn't it basically be:
```
for (Edge e1 : edges){
     for (Edge e2 : edges){
        if (intersects(e1, e2))  {
             return true
        }
    }
}
return false;
```

File: src/software/new_geom/util/intersects.cpp
Comment (garethellis0) [reply src/software/new_geom/util/intersects.cpp:19]: Ah, so will those eventually end up in this file? Or is there a reason for keeping some `intersects` methods in `util`?

File: src/software/ai/evaluation/indirect_chip.cpp
Comment (jonathanlew) [reply src/software/ai/evaluation/indirect_chip.cpp]: How about _centre_? But actually, I'm indifferent as well

File: src/software/ai/evaluation/indirect_chip.cpp
Comment (jonathanlew) [inline src/software/ai/evaluation/indirect_chip.cpp]: Out of scope, but I'd appreciate replacing this with `for each Triangle`
```suggestion
    for (Triangle t : triangles)
```

File: src/software/ai/evaluation/indirect_chip.cpp
Comment (jonathanlew) [inline src/software/ai/evaluation/indirect_chip.cpp]: Can probably use `for each` here too

**Review by jonathanlew — CHANGES_REQUESTED** [review]
Found a couple of improvements that are somewhat out of scope, but it seems overkill to make a new issue to fix them, so I'd appreciate if you could include them with this PR. Otherwise, looking much cleaner! :+1:

File: src/software/new_geom/convex_polygon.cpp
Comment (MathewMacDougall) [reply src/software/new_geom/convex_polygon.cpp]: Regarding the above pseudocode, just need to be careful about your edges touching at the vertices and whether `intersects` returns `true` in that case or not. Possible "interface complexity" instead of time complexity.

File: src/software/ai/evaluation/indirect_chip.cpp
Comment (scveloso) [reply src/software/ai/evaluation/indirect_chip.cpp]: Also indifferent

File: src/software/new_geom/util/intersects.cpp
Comment (scveloso) [reply src/software/new_geom/util/intersects.cpp:19]: I believe they will eventually be moved to this file once we take a look at refactoring ```geom/util``` - I'm not sure if @jsn0716 has other plans for it

File: src/software/new_geom/convex_polygon.cpp
Comment (scveloso) [reply src/software/new_geom/convex_polygon.cpp]: I was thinking of implementing the bentley ottoman algorithm to avoid having ```O(n^2)``` but that algorithm isn't trivial. If we're happy with your pseudocode then it should be fine given Mathew's comment - I can check in my next update

Comment (garethellis0) [reply src/software/new_geom/convex_polygon.cpp]: Yeah, I agree `O(n^2)` is not ideal, but I think for the vast majority of our use cases right now it'd be totally fine. Bentley-Ottoman would definitely be much better though, maybe just a file an issue so we don't forget to come back and change it!

File: src/software/ai/evaluation/indirect_chip.cpp
Comment (garethellis0) [reply src/software/ai/evaluation/indirect_chip.cpp]: Ok, let's change to `mean` then, since it looks like I'm the only one who's not indifferent here. Out-of-scope, I'm happy to make the change later if you don't want to do it here @scveloso, just let me know.

File: src/software/new_geom/util/intersects.cpp
Comment (jsn0716) [reply src/software/new_geom/util/intersects.cpp:19]: Yeah will probably try to group those remaining util functions like `intersects` into one PR, also `segment` has yet to be rewritten

File: src/software/new_geom/angle.h
Comment (jsn0716) [inline src/software/new_geom/angle.h]: Why was this change needed?

**Review by jsn0716 — COMMENTED** [review]
Looks good to me, not much to comment on beyond comments already given

File: src/software/new_geom/convex_polygon.cpp
Comment (scveloso) [reply src/software/new_geom/convex_polygon.cpp]: I tried to do a simple implementation of above pseudocode but since geom/util still has intersects and geom/util now depends on new geom shape hierarchy, we end up with a circular dependency and the build fails. I think I'm going to keep it as is for now and file an issue to fix it later if that's okay with you guys @garethellis0 @MathewMacDougall

File: src/software/new_geom/angle.h
Comment (scveloso) [reply src/software/new_geom/angle.h]: Leftover from what I was trying out previously. Removed in latest update

File: src/software/ai/evaluation/indirect_chip.cpp
Comment (scveloso) [reply src/software/ai/evaluation/indirect_chip.cpp]: Fixed in latest update

File: src/software/ai/evaluation/indirect_chip.cpp
Comment (scveloso) [reply src/software/ai/evaluation/indirect_chip.cpp]: Fixed in latest update

File: src/software/ai/evaluation/indirect_chip.cpp
Comment (scveloso) [reply src/software/ai/evaluation/indirect_chip.cpp]: Changed it to mean in latest update

File: src/software/new_geom/convex_polygon.cpp
Comment (garethellis0) [reply src/software/new_geom/convex_polygon.cpp]: Yeah ofc. Just file a ticket with everything you have so far, stuff discussed here, anything you've implemented so far, etc. so we don't lose info/work. 

Sorry about this, really didn't mean to hold up this PR. :grimacing:

**Review by garethellis0 — APPROVED** [review]
LGTM once that ticket for improving the stability of `isConvex` is filed.  :+1: :tada: - nice job, very happy to see this going in.

**Review by jsn0716 — APPROVED** [review]
LGTM!

File: src/software/new_geom/convex_polygon.cpp
Comment (scveloso) [reply src/software/new_geom/convex_polygon.cpp]: No problem, filed ticket #1140 for this

**Review by MathewMacDougall — APPROVED** [review]
:+1: Awesome stuff

**Review by jonathanlew — APPROVED** [review]
Very nice! :+1:

**Pull request merged**
